### ISSUE TITLE
Setting unFree for overriden, pinned packages

### ISSUE THREAD

**Issue by jokogr** [NixOS/nixpkgs#18527]
So far I was used to override some packages by using a future version of nixpkgs, e.g.:

```
pkgs: let
  inherit (pkgs) callPackage;
  fetchNixPkgs = { rev, sha256 }:
  pkgs.fetchFromGitHub {
    inherit sha256 rev;
    owner = "jokogr";
    repo = "nixpkgs";
  };
  pinnedPkgs = import (fetchNixPkgs {
    rev = "b4e2b6bc6aac50094dad02a01d34d0aaca7eb3f4";
    sha256 = "03c54sfnkrcp833wd9xq98lf5f4rq32mibpzlc1chypj52kmvgy8";
  }) {};
in {
  inherit (pinnedPkgs) texlive idea;
}
```

Although TeXLive is working as intended, pulling idea-ultimate shows the following error message:

```
error: Package ‘idea-ultimate-2016.2.4’ in ‘/nix/store/haawb90f29ick4frkfz3q9g9rxclb77n-nixpkgs-b4e2b6bc6aac50094dad02a01d34d0aaca7eb3f4-src/pkgs/applications/editors/idea/default.nix:33’ has an unfree license (‘unfree’), refusing to evaluate.
a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.nixpkgs/config.nix.

(use ‘--show-trace’ to show detailed location information)
```

Of course, I have already set `nixpkgs.config.allowUnfree = true;` in the configuration, but it seems to be ignored. Any ideas?

**Comment by jokogr** [comment]
I have even tried to do a nasty hack, by trying to override the package license:

```
  idea = pinnedPkgs.idea // {
  idea-ultimate = pinnedPkgs.lib.overrideDerivation pinnedPkgs.idea.idea-ultimate (attrs: attrs // {
    license = pinnedPkgs.stdenv.lib.licenses.asl20;
  });
```

The error message stays the same, though :confused:

**Comment by rasendubi** [comment]
I believe that's the same issue as https://github.com/NixOS/nixpkgs/issues/7541 (`meta` attribute is too strict).

**Comment by rasendubi** [comment]
Looked closely to the original issue: config is not passed to the `pinnedPkgs`.

The following works for me:
```nix
let pkgs = import <nixpkgs> {}; in
let
  inherit (pkgs) callPackage;
  fetchNixPkgs = { rev, sha256 }:
  pkgs.fetchFromGitHub {
    inherit sha256 rev;
    owner = "jokogr";
    repo = "nixpkgs";
  };
  pinnedPkgs = import (fetchNixPkgs {
    rev = "b4e2b6bc6aac50094dad02a01d34d0aaca7eb3f4";
    sha256 = "03c54sfnkrcp833wd9xq98lf5f4rq32mibpzlc1chypj52kmvgy8";
  }) {};
in {
  inherit (pinnedPkgs) texlive idea;
}
```
```
rasen@Larry ~ % nix-build try.nix -A idea.idea-ultimate
... works ok
rasen@Larry ~ % nix-env -f try.nix -iA idea.idea-ultimate
... works ok
rasen@Larry ~ % cat ~/.nixpkgs/config.nix
{
  allowUnfree = true;
  useChroot = true;
}
```
(I believe, that's mostly because I use nix-build/nix-env and not nixos-rebuild.)

If I understand correctly, you're using this config as part of you `nixos/configuration.nix`. If so, could you please post a minimal complete example, so I could reproduce the issue?

**Comment by abbradar** [comment]
If it's in `configuration.nix`, try:
```
  pinnedPkgs = import (fetchNixPkgs {
    rev = "b4e2b6bc6aac50094dad02a01d34d0aaca7eb3f4";
    sha256 = "03c54sfnkrcp833wd9xq98lf5f4rq32mibpzlc1chypj52kmvgy8";
  }) { inherit config; };
```
(notice `inherit config;`)

**Comment by jokogr** [comment]
@rasendubi, you may find my configuration @ https://git.joko.gr/joko/etc-nixos

Basically, just put `nixpkgs.config.packageOverrides = import ./pkgs.nix;` in your configuration.nix and then the contents I've pasted to a file pkgs.nix placed in the same directory.

@abbradar, to use your snippet I've changed my `pkgs.nix` accordingly:

```
{ config, pkgs, ... }:

let
  inherit (pkgs) callPackage;
  fetchNixPkgs = { rev, sha256 }:
  pkgs.fetchFromGitHub {
    inherit sha256 rev;
    owner = "jokogr";
    repo = "nixpkgs";
  };
  pinnedPkgs = import (fetchNixPkgs {
    rev = "b4e2b6bc6aac50094dad02a01d34d0aaca7eb3f4";
    sha256 = "03c54sfnkrcp833wd9xq98lf5f4rq32mibpzlc1chypj52kmvgy8";
  }) { inherit config; };
in {
  inherit (pinnedPkgs) texlive idea;
}
```

This created some load on my machine and after a while it still failed:

```
% sudo nixos-rebuild switch
building Nix...
building the system configuration...
Too many heap sections: Increase MAXHINCR or MAX_HEAP_SECTS
nix-instantiate killed by signal 6
```
I think by inheriting config, it goes into a never-ending loop of overriding packages.

**Comment by vcunat** [comment]
Well, if you use this snippet within your `packageOverrides`... but you don't need to inherit the whole config (you can e.g. purge `packageOverrides` from it or something).

**Comment by jokogr** [comment]
@vcunat could you make a clear example?

**Comment by vcunat** [comment]
@jokogr: by that you only added `config.config.nixpkgs...`, i.e. the nesting wasn't correct. It's a bit troublesome that `//` isn't "recursive". I'd probably use something like:
```nix
let # to fix highlighting
  import yourNixpkgsPath {
    config = lib.recursiveUpdate config.nixpkgs { config.packageOverrides = {}; }; 
  }
```

**Comment by jokogr** [comment]
@vcunat: I now get `error: attribute ‘nixpkgs’ missing` :confused:

**Comment by vcunat** [comment]
Fixed by 67e8392383, probably.

**Comment by AleXoundOS** [comment]
Not a solution but an alternative approach:
```nix
let
  unfreePredicate = (pkg: builtins.elem (builtins.parseDrvName pkg.pname).name
    [ "flashplayer-standalone" ]);
  unstable = import (builtins.fetchTarball {
    name = "nixos-unstable-2020-04-02";
    url = https://github.com/nixos/nixpkgs/archive/e20f308b8e53aad75e61ebbd211d6cdd92592b97.tar.gz;
    sha256 = "07b0mkd9gl8xk01mb3mb4bkj38nfxmfn3v14dx6zaph197mpx6w9";
  }) { config = { allowUnfreePredicate = unfreePredicate; }; };
in
{
  # ...
  environment.systemPackages = with pkgs; [
    unstable.flashplayer-standalone
  ];

}
```
